vmm/src/controller.rs

```diff
@@ -0,0 +1,1006 @@
+                    LoggerError::NeverInitialized(s) => (ErrorKind::User, s),
+                    _ => (ErrorKind::Internal, e.to_string()),
+                };
+                VmmActionError::Logger(kind, LoggerConfigError::FlushMetrics(error_contents))
+            })
+        }
+    }
+
+    /// Injects CTRL+ALT+DEL keystroke combo to the inner Vmm (if present).
+    #[cfg(target_arch = "x86_64")]
+    pub fn send_ctrl_alt_del(&mut self) -> UserResult {
+        if let Some(vmm) = self.vmm.as_mut() {
+            vmm.send_ctrl_alt_del()
+        } else {
+            // TODO: An error would prob be more informative.
+            Ok(())
+        }
+    }
+
+    /// Stops the inner Vmm (if present) and exits the process with the provided exit_code.
+    pub fn stop(&mut self, exit_code: i32) {
+        if let Some(vmm) = self.vmm.as_mut() {
+            // This currently exits the process.
+            vmm.stop(exit_code)
+        } else {
+            process::exit(exit_code)
+        }
+    }
+
+    /// Creates a new `VmmController`.
+    pub fn new(
+        api_shared_info: Arc<RwLock<InstanceInfo>>,
+        afi_event_fd: &EventFd,
+        seccomp_level: u32,
+    ) -> Result<Self> {
+        let device_configs = DeviceConfigs::new(
+            BlockDeviceConfigs::new(),
+            NetworkInterfaceConfigs::new(),
+            None,
+        );
+
+        let mut epoll_context = EpollContext::new()?;
+        epoll_context
+            .add_epollin_event(afi_event_fd, EpollDispatch::VmmActionRequest)
+            .expect("Cannot add vmm control_fd to epoll.");
+
+        Ok(VmmController {
+            device_configs,
+            epoll_context: Some(epoll_context),
+            guest_memory: None,
+            kernel_config: None,
+            vm_config: VmConfig::default(),
+            shared_info: api_shared_info,
+            vmm: None,
+            seccomp_level,
+            instance_initialized: false,
+        })
+    }
+
+    fn load_kernel(&mut self) -> std::result::Result<GuestAddress, StartMicrovmError> {
+        use StartMicrovmError::*;
+
+        // This is the easy way out of consuming the value of the kernel_cmdline.
+        let kernel_config = self.kernel_config.as_mut().ok_or(MissingKernelConfig)?;
+
+        // It is safe to unwrap because the VM memory was initialized before in vm.memory_init().
+        let guest_memory = self.guest_memory.as_ref().ok_or(GuestMemory(
+            memory_model::GuestMemoryError::MemoryNotInitialized,
+        ))?;
+
+        let entry_addr = kernel_loader::load_kernel(
+            guest_memory,
+            &mut kernel_config.kernel_file,
+            arch::get_kernel_start(),
+        )
+        .map_err(KernelLoader)?;
+
+        Ok(entry_addr)
+    }
+
+    /// Starts a microVM based on the current configuration.
+    pub fn start_microvm(&mut self) -> UserResult {
+        self.init_guest_memory()?;
+        let kernel_entry_addr = self.load_kernel()?;
```

Comment:
I cannot decide whether or not `load_kernel` should stay here and not in the `VmmBuilder`.
What is you reasoning behind keeping it here?
Shouldn the vmmbuilder be  the one executing low level boot protocols?
Also this makes both vmmbuilder and controller dependent on arch. I d like to see that dependency in only one place (vmmbuilder).